core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -827,159 +848,170 @@ public T next() {
             }
         }
 
+        public int getNextKeyIndex() {
+            return nextIndex;
+        }
+
         // the actual next()
         abstract T internalNext();
 
         /**
          * Advances next to higher entry.
+         * Return previous index
          */
-        void advance() {
-            if (next == Chunk.NONE)
+        Pair<ByteBuffer, Handle> advance() {
+
+            Object lastKey = lastKeyBeforeChunkDelete.get();
+            if (!(lastKey instanceof Chunk)) {
+                //Current chunk got released
+                if (isDescending) {
+                    hiInclusive = true;
+                    hi = (K)lastKeyBeforeChunkDelete.get();
+                }
+                else {
+                    loInclusive = true;
+                    lo = (K)lastKeyBeforeChunkDelete.get();
+                }
+                initChunk();
+                if (nextChunkIter == null) {
+                    throw new NoSuchElementException();
+                }
+                if (nextIndex == Chunk.NONE) {
+                    throw new ConcurrentModificationException();
+                }
+            }
+
+            if (nextIndex == Chunk.NONE) {
                 throw new NoSuchElementException();
-            findNextKeyInRange();
+            }
+
+
+            ByteBuffer bb = getKey(nextIndex, nextChunk);
+            Handle currentHandle = nextHandle;
+            setNextKey();
```

Comment:
this can sign out from viewers
do you need to check this?

Reply:
I dont understand you comment.
setNextKey can sigout and signin into a new chunk. why does this bother you?